### PR TITLE
[SPARK-18838][core] Add separate listener queues to LiveListenerBus.

### PR DESCRIPTION
**Pull request by vanzin**
This change modifies the live listener bus so that all listeners are
added to queues; each queue has its own thread to dispatch events,
making it possible to separate slow listeners from other more
performance-sensitive ones.

The public API has not changed - all listeners added with the existing
"addListener" method, which after this change mostly means all
user-defined listeners, end up in a default queue. Internally, there's
an API allowing listeners to be added to specific queues, and that API
is used to separate the internal Spark listeners into 3 categories:
application status listeners (e.g. UI), executor management (e.g. dynamic
allocation), and the event log.

The queueing logic, while abstracted away in a separate class, is kept
as much as possible hidden away from consumers. Aside from choosing their
queue, there's no code change needed to take advantage of queues.

Test coverage relies on existing tests; a few tests had to be tweaked
because they relied on `LiveListenerBus.postToAll` being synchronous,
and the change makes that method asynchronous. Other tests were simplified
not to use the asynchronous LiveListenerBus.

### CODE REVIEW

**Comment by vanzin** [comment]
(Nevermind the test failures, I killed the obsolete builds.)

**Comment by vanzin** [comment]
retest this please

**Comment by vanzin** [comment]
I think this is ready now, pinging some people: @squito @JoshRosen @cloud-fan

**Comment by ajbozarth** [comment]
I did a read-through of your code and didn't find any glaring issues, but I'd like to go through it a bit more before passing judgment. Overall a nice new feature though.

File: core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala [hunk 1]
```diff
@@ -217,7 +217,7 @@ private[spark] class ExecutorAllocationManager(
    * the scheduling task.
    */
   def start(): Unit = {
-    listenerBus.addListener(listener)
+    listenerBus.addToQueue(listener, LiveListenerBus.EXECUTOR_MGMT_QUEUE)
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: will we have more queues in the future? If it's not too many, maybe we can use different methods, like `addToMGMTQueue`

File: core/src/main/scala/org/apache/spark/scheduler/AsyncEventQueue.scala [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+
+  override protected def getTimer(listener: SparkListenerInterface): Option[Timer] = {
+    metrics.getTimerForListenerClass(listener.getClass.asSubclass(classOf[SparkListenerInterface]))
+  }
+
+  /**
+   * Start an asynchronous thread to dispatch events to the underlying listeners.
+   *
+   * @param sc Used to stop the SparkContext in case the a listener fails.
+   * @param metrics Used to report listener performance metrics.
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/scheduler/AsyncEventQueue.scala]: there is no `metrics` param.

File: core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala [hunk 3]
```diff
@@ -65,53 +60,61 @@ private[spark] class LiveListenerBus(conf: SparkConf) extends SparkListenerBus {
...
-              postToAll(event)
-            } finally {
-              timerContext.stop()
-            }
-          } finally {
-            self.synchronized {
-              processingEvent = false
-            }
-          }
+  private val queues = new CopyOnWriteArrayList[AsyncEventQueue]()
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala]: can we have something like
```
private val defaultQueue = ...
private val executorManagementQueue = ...
...
```

I think the number of queues will be small.

Comment (vanzin) [reply core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala]: I can add specific methods for each internal queue; but I'd like to keep the internal management of queues more generic. One of the ideas in #18253 was to allow filtering of which events are enqueued at all (e.g. don't enqueue `SparkListenerBlockUpdated` because it's not written to the event logs, reducing the load on the event log queue). Leaving the internal management more generic would allow that to be more easily / cleanly implemented later (instead of "addToEventLogQueue", you'd have a "addCustomQueue" method with a subclass of `AsyncEventQueue`).

File: core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala [hunk 4]
```diff
@@ -65,53 +60,76 @@ private[spark] class LiveListenerBus(conf: SparkConf) extends SparkListenerBus {
...
+    }
+
+    queues.asScala.find(_.name == queue) match {
+      case Some(queue) =>
+        queue.addListener(listener)
+
+      case None =>
+        val newQueue = new AsyncEventQueue(queue, conf, metrics)
+        newQueue.addListener(listener)
+        if (started.get() && !stopped.get()) {
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala]: `stopped.get` is always false here.

File: core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala [hunk 5]
```diff
@@ -65,53 +60,76 @@ private[spark] class LiveListenerBus(conf: SparkConf) extends SparkListenerBus {
...
+  }
+
+  /**
+   * Add a listener to a specific queue, creating a new queue if needed. Queues are independent
+   * of each other (each one uses a separate thread for delivering events), allowing slower
+   * listeners to be somewhat isolated from others.
+   */
+  private def addToQueue(listener: SparkListenerInterface, queue: String): Unit = synchronized {
+    if (stopped.get()) {
+      throw new IllegalStateException("LiveListenerBus is stopped.")
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala]: shall we also check `!started.get`?

File: core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala [hunk 6]
```diff
@@ -65,53 +60,76 @@ private[spark] class LiveListenerBus(conf: SparkConf) extends SparkListenerBus {
...
+  def removeListener(listener: SparkListenerInterface): Unit = synchronized {
+    // Remove listener from all queues it was added to, and stop queues that have become empty.
+    queues.asScala
+      .filter { queue =>
+        queue.removeListener(listener)
+        queue.listeners.isEmpty()
+      }
+      .foreach { toRemove =>
+        if (started.get() && !stopped.get()) {
+          toRemove.stop()
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala]: this logic is a little weird, why we stop the queue when there is no listeners? is it possible that we may add listeners back to this queue later?

File: core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala [hunk 7]
```diff
@@ -65,53 +60,76 @@ private[spark] class LiveListenerBus(conf: SparkConf) extends SparkListenerBus {
...
+        }
+        queues.remove(toRemove)
+      }
+  }
+
+  /** Post an event to all queues. */
+  def post(event: SparkListenerEvent): Unit = {
+    if (!stopped.get()) {
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala]: shall we add an assert that this bus is started?

File: core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala (hunk 6 above)
Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala]: oh I see your point, we can re-create and start this queue later if needed. Is it really necessary? Can't we keep all the queues active all the time?

File: core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala (hunk 5 above)
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala]: No. It's ok and actually expected to add listeners before the bus is started.

File: core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala (hunk 6 above)
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala]: That would waste resources in the normal case.

The default configuration would never add anything to the default and event log queues, so you'd have two threads dispatching events to no listeners.

Stopping the queue when it empties, while a rare case, handles when users add listeners and later remove them, potentially leaving the default queue empty. Singling out the default queue here just complicates the code.

File: core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala (hunk 7 above)
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala]: No, as the class javadoc says, it's ok to send events before the bus is started.

File: core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala [hunk 8]
```diff
@@ -39,20 +41,13 @@ import org.apache.spark.util.Utils
...
  * is stopped when `stop()` is called, and it will drop further events after stopping.
  */
-private[spark] class LiveListenerBus(conf: SparkConf) extends SparkListenerBus {
-
-  self =>
+private[spark] class LiveListenerBus(conf: SparkConf) {
 
   import LiveListenerBus._
 
   private var sparkContext: SparkContext = _
```
Comment (tgravescs) [inline core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala]: maybe I missed it but I don't see this being set anywhere now

Comment (vanzin) [reply core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala]: Ah, this should be set in `start()`, probably deleted in one of the code moves.

File: core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala [hunk 9]
```diff
@@ -173,80 +159,62 @@ private[spark] class LiveListenerBus(conf: SparkConf) extends SparkListenerBus {
...
+      queues.clear()
     }
-    logTrace(s"Dropping event $event")
   }
+
+  private[spark] def findListenersByClass[T <: SparkListenerInterface : ClassTag](): Seq[T] = {
+    queues.asScala.flatMap { queue => queue.findListenersByClass[T]() }
+  }
+
+  private[spark] def listeners: JList[SparkListenerInterface] = {
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala]: where do we use this method?

File: core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala [hunk 10]
```diff
@@ -173,80 +159,62 @@ private[spark] class LiveListenerBus(conf: SparkConf) extends SparkListenerBus {
...
-        "This likely means one of the SparkListeners is too slow and cannot keep up with " +
-        "the rate at which tasks are being started by the scheduler.")
+    synchronized {
+      queues.asScala.foreach(_.stop())
+      queues.clear()
     }
-    logTrace(s"Dropping event $event")
   }
+
+  private[spark] def findListenersByClass[T <: SparkListenerInterface : ClassTag](): Seq[T] = {
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala]: let's add a comment to say it testing only.

File: core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala [hunk 11]
```diff
@@ -173,80 +159,62 @@ private[spark] class LiveListenerBus(conf: SparkConf) extends SparkListenerBus {
...
   // Allows for Context to check whether stop() call is made within listener thread
   val withinListenerThread: DynamicVariable[Boolean] = new DynamicVariable[Boolean](false)
 
-  /** The thread name of Spark listener bus */
-  val name = "SparkListenerBus"
+  private[scheduler] val SHARED_QUEUE = "shared"
+
+  private[scheduler] val APP_STATUS_QUEUE = "appStatus"
+
+  private[scheduler] val EXECUTOR_MGMT_QUEUE = "executorMgmt"
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala]: `MGMT` is hard to understand, how about `MANAGEMENT`?

File: core/src/test/scala/org/apache/spark/scheduler/SparkListenerSuite.scala [hunk 12]
```diff
@@ -42,59 +44,65 @@ class SparkListenerSuite extends SparkFunSuite with LocalSparkContext with Match
...
+    assert(queueSize(bus) === 0)
+    assert(eventProcessingTimeCount(bus) === 5)
 
     // After listener bus has stopped, posting events should not increment counter
     bus.stop()
     (1 to 5).foreach { _ => bus.post(SparkListenerJobEnd(0, jobCompletionTime, JobSucceeded)) }
     assert(counter.count === 5)
-    assert(bus.metrics.numEventsPosted.getCount === 5)
-
-    // Make sure per-listener-class timers were created:
```
Comment (cloud-fan) [inline core/src/test/scala/org/apache/spark/scheduler/SparkListenerSuite.scala]: do we still have this?

Comment (cloud-fan) [reply core/src/test/scala/org/apache/spark/scheduler/SparkListenerSuite.scala]: oh seems we don't need it anymore.

File: streaming/src/test/scala/org/apache/spark/streaming/StreamingContextSuite.scala [hunk 13]
```diff
@@ -575,8 +575,6 @@ class StreamingContextSuite extends SparkFunSuite with BeforeAndAfter with TimeL
 
   test("getActive and getActiveOrCreate") {
     require(StreamingContext.getActive().isEmpty, "context exists from before")
-    sc = new SparkContext(conf)
```
Comment (cloud-fan) [inline streaming/src/test/scala/org/apache/spark/streaming/StreamingContextSuite.scala]: why this change?

File: core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala [hunk 14]
```diff
@@ -39,20 +41,13 @@ import org.apache.spark.util.Utils
  * has started will events be actually propagated to all attached listeners. This listener bus
  * is stopped when `stop()` is called, and it will drop further events after stopping.
  */
-private[spark] class LiveListenerBus(conf: SparkConf) extends SparkListenerBus {
-
-  self =>
+private[spark] class LiveListenerBus(conf: SparkConf) {
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala]: is it still valid to call `LiveListenerBus.postToAll`? If not, shall we overwrite `postToAll` and throw exception?

Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala]: also `LiveListenerBus.addListener` seems not valid too.

**Comment by cloud-fan** [comment]
LGTM except some minor comments

File: core/src/main/scala/org/apache/spark/scheduler/AsyncEventQueue.scala [hunk 15]
```diff
@@ -0,0 +1,196 @@
...
+   * Start an asynchronous thread to dispatch events to the underlying listeners.
+   *
+   * @param sc Used to stop the SparkContext in case the async dispatcher fails.
+   */
+  private[scheduler] def start(sc: SparkContext): Unit = {
+    if (started.compareAndSet(false, true)) {
+      this.sc = sc
+      dispatchThread.start()
+    } else {
+      throw new IllegalStateException(s"$name already started!")
```
Comment (jerryshao) [inline core/src/main/scala/org/apache/spark/scheduler/AsyncEventQueue.scala]: Is it better to be a no-op here, or a warning log instead?

File: core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala (hunk 9 above)
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala]: Used by tests.

File: core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala (hunk 11 above)
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala]: That's a pretty common abbreviation for "management", but sure.

File: streaming/src/test/scala/org/apache/spark/streaming/StreamingContextSuite.scala (hunk 13 above)
Comment (vanzin) [reply streaming/src/test/scala/org/apache/spark/streaming/StreamingContextSuite.scala]: Because the test stops the underlying streaming context and thus the spark context, and you can't add listeners to the bus after that. Thus, it was failing.

File: core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala (hunk 14 above)
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala]: `LiveListenerBus.postToAll` doesn't exist anymore, and neither does `LiveListenerBus.addListener`. But I'm not sure what you're asking.

File: core/src/main/scala/org/apache/spark/scheduler/AsyncEventQueue.scala (hunk 15 above)
Comment (squito) [reply core/src/main/scala/org/apache/spark/scheduler/AsyncEventQueue.scala]: seems fine to me -- its only called in LiveListenerBus, where we guarantee this is true.  seems better to fail-fast if its messed up

**Comment by squito** [comment]
lgtm

do you know how much better this makes it?  Eg., if we had an existing case where things go haywire in dynamic allocation because of this -- we could see if after this change, the only dropped events are in eventLog, hopefully.  Seems like the right change regardless, but it would be good to know the effect.

**Comment by vanzin** [comment]
I don't really have a readily available cluster that can easily put that much pressure on the listener bus; I tried with the one I have and monitored the listener bus metrics, and found pretty much what I expected (slightly more latency in the event log queue, but not enough in my tests to actually cause event drops).

File: core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala (hunk 14 above)
Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala]: oh sorry I missed that, `LiveListenerBus` doesn't extend `SparkListenerBus` anymore.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by dongjoon-hyun** [comment]
Hi, @vanzin and @cloud-fan .
This seems to break the master branch. Could you take a look this?
- https://amplab.cs.berkeley.edu/jenkins/view/Spark%20QA%20Test%20(Dashboard)/job/spark-master-test-sbt-hadoop-2.7/3508/

**Comment by vanzin** [comment]
Following assert is failing in some cases, let me fix it and send a follow up PR:

```
    val listeners = sc.listenerBus.findListenersByClass[SaveExecutorInfo]
    assert(listeners.size === 1)
```
